# Phase 09: Robustness and Generalization

Executed initialization and feasibility audit only. No model training, prediction generation, OOF generation, hyperparameter search, or formal statistical analysis is performed.

## 1. Stage purpose and evidence boundary

The main evidence is Phase 03 Primary without performance features. Frozen Phase 03 folds and Phase 04A/04B/06/07/08 interfaces are read-only. The planned workstreams are missing-modality robustness, selected-model LOSO, and subject-level stability.

In [1]:
from pathlib import Path
import json, os, platform, subprocess, sys
PHASE09_ROOT = Path.cwd().resolve()
assert PHASE09_ROOT.name == 'phase_09_robustness_and_generalization'
sys.path.insert(0, str(PHASE09_ROOT / 'scripts'))
from initialize_phase09 import run_initialization, REQUIRED_ARTIFACTS, REQUIRED_DIRECTORIES
environment = {'python': sys.version, 'platform': platform.platform(), 'working_directory': os.getcwd()}
environment

{'python': '3.12.7 | packaged by Anaconda, Inc. | (main, Oct  4 2024, 13:17:27) [MSC v.1929 64 bit (AMD64)]',
 'platform': 'Windows-11-10.0.26200-SP0',
 'working_directory': 'E:\\hdc-vr-pilot\\experiments\\phase_09_robustness_and_generalization'}

## 2. Run read-only initialization and upstream freeze validation

In [2]:
summary = run_initialization()
upstream = json.loads((PHASE09_ROOT / 'audits/phase09_upstream_freeze_audit.json').read_text(encoding='utf-8'))
{'initialization': summary, 'phase_freeze_interfaces': upstream['interface_results'], 'actual_freeze_sha256': upstream['actual_freeze_sha256'], 'status': upstream['status']}

{'initialization': {'phase': '09',
  'phase_name': 'Robustness and Generalization',
  'status': 'PENDING_CONTRACT_FREEZE',
  'input_status': 'PASS',
  'upstream_status': 'PASS',
  'modality_status': 'PASS',
  'loso_status': 'PASS',
  'generalization_scope_status': 'PASS',
  'upstream_sources_unchanged': True,
  'ready_for_contract_freeze_pre_notebook': True,
  'ready_for_modeling': False,
  'model_training_executed': False,
  'loso_predictions_generated': False,
  'missing_modality_predictions_generated': False},
 'phase_freeze_interfaces': {'phase04a': True,
  'phase04b': True,
  'phase06': True,
  'phase07': True,
  'phase08': True},
 'actual_freeze_sha256': {'phase04a': '34ea8100d9406f9701750a441aa6537323c28bcdb194cb3fd3645c4f7de4a2e1',
  'phase04b': 'e2c88b1139a50aab6d47b6477c7bceff74f8443095f9d039ea9af84b715ee790',
  'phase06': 'cd94acdf0bcb463b5f48a7c84110a23059899b6b5bf156131694827d7d7bde66',
  'phase07': '8569b48a8210f0ca1316d5a140d292edb892e2a556ea13ee299e9b97699af492',
  'pha

## 3. Primary data preflight

In [3]:
input_audit = json.loads((PHASE09_ROOT / 'audits/phase09_input_and_fold_audit.json').read_text(encoding='utf-8'))
{'actual': input_audit['actual'], 'checks': input_audit['checks'], 'status': input_audit['status']}

{'actual': {'rows': 419,
  'subjects': 35,
  'primary_features': 1176,
  'unique_run_keys': 419,
  'target_class_values': [0, 1, 2, 3],
  'target_score_values': [1.0, 2.0, 3.0, 4.0],
  'outer_folds': 5,
  'primary_sha256': '0a2aef89c01b43c3a4e5afe40b96797784627665490982731b2541f96b45fc44',
  'fold_sha256': 'e4dc943af21851bace49345f6336f9c88f82613ca3a26b3c433efe7dfb041f6f'},
 'checks': {'rows': True,
  'subjects': True,
  'primary_features': True,
  'unique_run_keys': True,
  'target_class_values': True,
  'target_score_values': True,
  'targets_complete': True,
  'manifest_features_unique': True,
  'manifest_equals_csv_feature_universe': True,
  'fold_rows_and_run_keys': True,
  'fold_alignment': True,
  'frozen_outer_folds': True,
  'outer_subject_isolation': True,
  'primary_checksum': True,
  'fold_checksum': True},
 'status': 'PASS'}

## 4. Frozen five-fold verification

In [4]:
{'fold_sha256': input_audit['actual']['fold_sha256'], 'checksum_pass': input_audit['checks']['fold_checksum'], 'outer_folds': input_audit['actual']['outer_folds'], 'fold_alignment': input_audit['checks']['fold_alignment'], 'subject_isolation': input_audit['checks']['outer_subject_isolation'], 'fold_details': input_audit['fold_details']}

{'fold_sha256': 'e4dc943af21851bace49345f6336f9c88f82613ca3a26b3c433efe7dfb041f6f',
 'checksum_pass': True,
 'outer_folds': 5,
 'fold_alignment': True,
 'subject_isolation': True,
 'fold_details': [{'outer_fold': 1,
   'train_subjects': 28,
   'test_subjects': 7,
   'subject_overlap': 0,
   'train_rows': 335,
   'test_rows': 84},
  {'outer_fold': 2,
   'train_subjects': 28,
   'test_subjects': 7,
   'subject_overlap': 0,
   'train_rows': 335,
   'test_rows': 84},
  {'outer_fold': 3,
   'train_subjects': 28,
   'test_subjects': 7,
   'subject_overlap': 0,
   'train_rows': 335,
   'test_rows': 84},
  {'outer_fold': 4,
   'train_subjects': 28,
   'test_subjects': 7,
   'subject_overlap': 0,
   'train_rows': 335,
   'test_rows': 84},
  {'outer_fold': 5,
   'train_subjects': 28,
   'test_subjects': 7,
   'subject_overlap': 0,
   'train_rows': 336,
   'test_rows': 83}]}

## 5. Five-modality feature counts and coverage

In [5]:
modality = json.loads((PHASE09_ROOT / 'audits/phase09_modality_coverage_audit.json').read_text(encoding='utf-8'))
{'counts': modality['modality_counts'], 'union_count': modality['union_count'], 'overlap_count': modality['overlap_count'], 'checks': modality['checks'], 'status': modality['status']}

{'counts': {'physiological_features': 233,
  'eye_tracking_features': 416,
  'head_movement_features': 159,
  'flight_parameter_features': 326,
  'body_movement_features': 42},
 'union_count': 1176,
 'overlap_count': 0,
 'checks': {'five_modalities': True,
  'declared_counts_match': True,
  'features_unique_within_modalities': True,
  'modalities_mutually_exclusive': True,
  'union_count_1176': True,
  'union_equals_primary_manifest': True,
  'all_missing_conditions_nonempty': True,
  'traditional_pipeline_feature_capacity': True,
  'hdc_feature_k_capacity': True},
 'status': 'PASS'}

## 6. Missing-modality condition feasibility

In [6]:
missing_plan = json.loads((PHASE09_ROOT / 'configs/phase09_missing_modality_plan.json').read_text(encoding='utf-8'))
{'conditions': missing_plan['conditions'], 'protocols_must_remain_separate': missing_plan['protocols_requiring_separate_contracts'], 'hdc_feasibility': missing_plan['hdc_feasibility'], 'traditional_feasibility': missing_plan['traditional_feasibility'], 'training_executed': missing_plan['training_executed'], 'predictions_generated': missing_plan['predictions_generated']}

{'conditions': [{'condition': 'full_primary',
   'feature_count': 1176,
   'removed_modality': None},
  {'condition': 'missing_physiological',
   'removed_modality': 'physiological_features',
   'removed_feature_count': 233,
   'feature_count': 943},
  {'condition': 'missing_eye_tracking',
   'removed_modality': 'eye_tracking_features',
   'removed_feature_count': 416,
   'feature_count': 760},
  {'condition': 'missing_head_movement',
   'removed_modality': 'head_movement_features',
   'removed_feature_count': 159,
   'feature_count': 1017},
  {'condition': 'missing_flight_parameter',
   'removed_modality': 'flight_parameter_features',
   'removed_feature_count': 326,
   'feature_count': 850},
  {'condition': 'missing_body_movement',
   'removed_modality': 'body_movement_features',
   'removed_feature_count': 42,
   'feature_count': 1134}],
 'protocols_must_remain_separate': {'retrain_without_modality': 'Remove the frozen modality before fold-local preprocessing, fit a new model using 

## 7. Deterministic 35-fold LOSO feasibility

In [7]:
loso = json.loads((PHASE09_ROOT / 'audits/phase09_loso_feasibility_audit.json').read_text(encoding='utf-8'))
loso_manifest = json.loads((PHASE09_ROOT / 'manifests/phase09_loso_feasibility_manifest.json').read_text(encoding='utf-8'))
{'split_count': loso['split_count'], 'checks': loso['checks'], 'first_split': loso_manifest['splits'][0], 'last_split': loso_manifest['splits'][-1], 'status': loso['status']}

{'split_count': 35,
 'checks': {'deterministic_subject_order': True,
  'split_count_35': True,
  'one_test_subject_each': True,
  'thirty_four_training_subjects_each': True,
  'subject_isolation': True,
  'run_key_isolation': True,
  'nonempty_splits': True,
  'unique_run_keys': True,
  'training_four_class_coverage': True,
  'test_four_class_coverage': True},
 'first_split': {'fold_id': 'loso_sub-cp003',
  'test_subject': 'sub-cp003',
  'train_subject_count': 34,
  'test_subject_count': 1,
  'train_rows': 407,
  'test_rows': 12,
  'subject_overlap': 0,
  'train_target_class_values': [0, 1, 2, 3],
  'test_target_class_values': [0, 1, 2, 3],
  'train_four_class_coverage': True,
  'test_four_class_coverage': True,
  'train_run_keys_unique': True,
  'test_run_keys_unique': True,
  'run_key_overlap': 0,
  'nonempty_train_and_test': True},
 'last_split': {'fold_id': 'loso_sub-cp043',
  'test_subject': 'sub-cp043',
  'train_subject_count': 34,
  'test_subject_count': 1,
  'train_rows': 407,


## 8. Subject run counts and target coverage

In [8]:
{'subject_run_counts': loso['subject_run_counts'], 'subject_target_coverage': loso['subject_target_coverage'], 'empty_training_sets': loso['empty_training_sets'], 'empty_test_sets': loso['empty_test_sets']}

{'subject_run_counts': {'sub-cp003': 12,
  'sub-cp004': 12,
  'sub-cp005': 12,
  'sub-cp006': 12,
  'sub-cp008': 12,
  'sub-cp009': 12,
  'sub-cp011': 12,
  'sub-cp012': 12,
  'sub-cp013': 12,
  'sub-cp014': 12,
  'sub-cp015': 12,
  'sub-cp016': 12,
  'sub-cp017': 12,
  'sub-cp018': 12,
  'sub-cp019': 12,
  'sub-cp020': 12,
  'sub-cp022': 12,
  'sub-cp023': 12,
  'sub-cp024': 12,
  'sub-cp025': 12,
  'sub-cp026': 12,
  'sub-cp027': 12,
  'sub-cp028': 12,
  'sub-cp029': 12,
  'sub-cp030': 11,
  'sub-cp031': 12,
  'sub-cp032': 12,
  'sub-cp033': 12,
  'sub-cp035': 12,
  'sub-cp036': 12,
  'sub-cp037': 12,
  'sub-cp038': 12,
  'sub-cp039': 12,
  'sub-cp042': 12,
  'sub-cp043': 12},
 'subject_target_coverage': {'sub-cp003': [0, 1, 2, 3],
  'sub-cp004': [0, 1, 2, 3],
  'sub-cp005': [0, 1, 2, 3],
  'sub-cp006': [0, 1, 2, 3],
  'sub-cp008': [0, 1, 2, 3],
  'sub-cp009': [0, 1, 2, 3],
  'sub-cp011': [0, 1, 2, 3],
  'sub-cp012': [0, 1, 2, 3],
  'sub-cp013': [0, 1, 2, 3],
  'sub-cp014': [0, 1, 2,

## 9. Generalization-scope limitations

In [9]:
scope = json.loads((PHASE09_ROOT / 'configs/phase09_generalization_scope.json').read_text(encoding='utf-8'))
scope_audit = json.loads((PHASE09_ROOT / 'audits/phase09_generalization_scope_audit.json').read_text(encoding='utf-8'))
{'subject_generalization': scope['subject_generalization'], 'missing_modality_robustness': scope['missing_modality_robustness'], 'unseen_session_generalization': scope['unseen_session_generalization'], 'unseen_scenario_generalization': scope['unseen_scenario_generalization'], 'task_template_generalization': scope['task_template_generalization'], 'route_configuration_generalization': scope['route_configuration_generalization'], 'flight_claim': scope['flight_generalizable_behavior_claim'], 'checks': scope_audit['checks']}

{'subject_generalization': 'FEASIBLE_VIA_LOSO',
 'missing_modality_robustness': 'FEASIBLE_PENDING_CONTRACT',
 'unseen_session_generalization': 'NOT_FEASIBLE_DUE_TO_METADATA',
 'unseen_scenario_generalization': 'NOT_FEASIBLE_DUE_TO_METADATA',
 'task_template_generalization': 'NOT_FEASIBLE_DUE_TO_METADATA',
 'route_configuration_generalization': 'NOT_FEASIBLE_DUE_TO_METADATA',
 'flight_claim': 'INCONCLUSIVE_DUE_TO_METADATA',
 'checks': {'phase08_handoff_limits_match': True,
  'session_equals_subject_partition': True,
  'required_grouping_metadata_absent': True,
  'difficulty_not_used_as_scenario': True}}

## 10. Selected frozen model interfaces

In [10]:
interfaces = json.loads((PHASE09_ROOT / 'configs/phase09_selected_model_interfaces.json').read_text(encoding='utf-8'))
{'traditional_classification': {'model': interfaces['traditional_classification']['model'], 'interface_pass': interfaces['traditional_classification']['interface_pass'], 'fold_specific_parameters': interfaces['traditional_classification']['fold_specific_parameters']}, 'traditional_regression': {'model': interfaces['traditional_regression']['model'], 'interface_pass': interfaces['traditional_regression']['interface_pass'], 'fold_specific_parameters': interfaces['traditional_regression']['fold_specific_parameters']}, 'hdc_classification': {'model': interfaces['hdc_classification']['selected_variant_name'], 'dimension': interfaces['hdc_classification']['selected_fixed_dimension'], 'interface_pass': interfaces['hdc_classification']['interface_pass']}, 'hdc_regression': {'model': interfaces['hdc_regression']['selected_regression_head'], 'dimension': interfaces['hdc_regression']['selected_fixed_dimension'], 'interface_pass': interfaces['hdc_regression']['interface_pass']}, 'policy': interfaces['policy']}

{'traditional_classification': {'model': 'Gradient Boosting',
  'interface_pass': True,
  'fold_specific_parameters': {'1': {'classifier__learning_rate': 0.1,
    'classifier__max_depth': 2,
    'classifier__n_estimators': 200,
    'selector__k': 200,
    'effective_selected_k': 200},
   '2': {'classifier__learning_rate': 0.05,
    'classifier__max_depth': 2,
    'classifier__n_estimators': 100,
    'selector__k': 200,
    'effective_selected_k': 200},
   '3': {'classifier__learning_rate': 0.05,
    'classifier__max_depth': 3,
    'classifier__n_estimators': 200,
    'selector__k': 200,
    'effective_selected_k': 200},
   '4': {'classifier__learning_rate': 0.05,
    'classifier__max_depth': 2,
    'classifier__n_estimators': 100,
    'selector__k': 200,
    'effective_selected_k': 200},
   '5': {'classifier__learning_rate': 0.1,
    'classifier__max_depth': 3,
    'classifier__n_estimators': 200,
    'selector__k': 200,
    'effective_selected_k': 200}}},
 'traditional_regression': {'

## 11. Initialization artifact audit and static tests

In [11]:
pre_persistence_missing_dirs = [p for p in REQUIRED_DIRECTORIES if not (PHASE09_ROOT / p).is_dir()]
deferred_self_audits = {'audits/phase09_notebook_persistence_audit.json', 'audits/phase09_initialization_artifact_audit.json'}
pre_persistence_missing_artifacts = [p for p in REQUIRED_ARTIFACTS if p not in deferred_self_audits and not (PHASE09_ROOT / p).is_file()]
test_process = subprocess.run([sys.executable, '-m', 'unittest', 'discover', '-s', str(PHASE09_ROOT / 'tests'), '-p', 'test_phase09_initialization.py', '-v'], cwd=PHASE09_ROOT, capture_output=True, text=True, check=False)
assert test_process.returncode == 0, test_process.stdout + test_process.stderr
{'pre_persistence_directory_audit': 'PASS' if not pre_persistence_missing_dirs else 'FAIL', 'pre_persistence_artifact_audit': 'PASS' if not pre_persistence_missing_artifacts else 'FAIL', 'missing_directories': pre_persistence_missing_dirs, 'missing_artifacts': pre_persistence_missing_artifacts, 'static_tests': 'PASS', 'test_details': (test_process.stdout + test_process.stderr)[-2500:], 'external_notebook_persistence_audit': 'RUN_AFTER_NOTEBOOK_SAVE'}

{'pre_persistence_directory_audit': 'PASS',
 'pre_persistence_artifact_audit': 'PASS',
 'missing_directories': [],
 'missing_artifacts': [],
 'static_tests': 'PASS',
 'test_details': 'test_all_upstream_freeze_interfaces (test_phase09_initialization.Phase09InitializationTests.test_all_upstream_freeze_interfaces) ... ok\ntest_input_and_frozen_fold (test_phase09_initialization.Phase09InitializationTests.test_input_and_frozen_fold) ... ok\ntest_loso_feasibility (test_phase09_initialization.Phase09InitializationTests.test_loso_feasibility) ... ok\ntest_modality_partition (test_phase09_initialization.Phase09InitializationTests.test_modality_partition) ... ok\ntest_scope_and_prohibited_operations (test_phase09_initialization.Phase09InitializationTests.test_scope_and_prohibited_operations) ... ok\n\n----------------------------------------------------------------------\nRan 5 tests in 0.554s\n\nOK\n',
 'external_notebook_persistence_audit': 'RUN_AFTER_NOTEBOOK_SAVE'}

## 12. Contract Freeze entry gate

In [12]:
contract = json.loads((PHASE09_ROOT / 'configs/phase09_experiment_contract_draft.json').read_text(encoding='utf-8'))
statistical_plan = json.loads((PHASE09_ROOT / 'configs/phase09_statistical_plan.json').read_text(encoding='utf-8'))
{'contract_status': contract['status'], 'next_action': contract['next_action'], 'subject_level_stability_plan': statistical_plan['planned_analyses'], 'training_authorized': contract['training_authorized'], 'prediction_generation_authorized': contract['prediction_generation_authorized'], 'formal_statistics_authorized': contract['formal_statistics_authorized'], 'ready_for_modeling': False}

{'contract_status': 'PENDING_CONTRACT_FREEZE',
 'next_action': 'Freeze the Phase 09 experiment contract in a separate step before modeling.',
 'subject_level_stability_plan': ['between-subject performance distribution',
  'worst-subject diagnosis',
  'subject-ranking stability',
  'subject-level bootstrap confidence intervals'],
 'training_authorized': False,
 'prediction_generation_authorized': False,
 'formal_statistics_authorized': False,
 'ready_for_modeling': False}

## Phase Validation Summary

**VERIFIED:** Primary input/folds, upstream frozen interfaces, five-modality partition, missing-modality computational feasibility, 35 deterministic LOSO splits, subject target coverage, and metadata-based scope limits.

**NOT VERIFIED:** Performance after modality loss or under LOSO; no such experiment was run.

**WARNINGS:** LOSO supports only subject-generalization claims. Flight generalizable behavior remains `INCONCLUSIVE_DUE_TO_METADATA`.

**KEY RESULTS:** Initialization only; no model metrics or predictions.

**OUTPUT FILES:** Draft configs, manifests, audits, this executed notebook, README, and notes.

**NEXT PHASE REQUIREMENTS:** Separate Phase 09 Contract Freeze before any modeling.

# Phase 09 Contract Freeze

This appended section freezes the Phase 09 evidence, model, missing-modality, LOSO, aggregation, and statistical contracts. It performs no training or prediction generation.

## 13. Contract Freeze summary

In [1]:
from pathlib import Path
import json, subprocess, sys
PHASE09_ROOT = Path.cwd().resolve()
sys.path.insert(0, str(PHASE09_ROOT / 'scripts'))
from freeze_phase09_contract import run_freeze
contract_summary = run_freeze()
contract_summary

{'contract_freeze_audit': 'PASS',
 'status': 'PENDING_NOTEBOOK_PERSISTENCE',
 'authorized_training_runs': 720,
 'missing_modality_runs': 300,
 'loso_runs': 420,
 'duplicate_run_identifiers': 0,
 'test_time_missingness': 'NOT_FEASIBLE_DUE_TO_CHECKPOINT_INTERFACE',
 'upstream_files_modified': 0,
 'training_executed': False,
 'predictions_generated': False}

## 14. Frozen selected-model interfaces

In [2]:
frozen = json.loads((PHASE09_ROOT / 'configs/phase09_frozen_contract.json').read_text(encoding='utf-8'))
frozen['selected_model_interfaces']

{'hdc_classification': {'model': 'HDC+OnlineHD Hybrid',
  'dimension': 5000,
  'levels': 51,
  'feature_k': 50,
  'seeds': [42, 43, 44, 45, 46],
  'fold_specific_structures': [{'outer_fold': 1,
    'selected_structure_json': '{"centroids_per_class":3,"epochs":1,"learning_rate":0.1,"margin_threshold":0.0}'},
   {'outer_fold': 2,
    'selected_structure_json': '{"centroids_per_class":2,"epochs":1,"learning_rate":0.05,"margin_threshold":0.0}'},
   {'outer_fold': 3,
    'selected_structure_json': '{"centroids_per_class":2,"epochs":3,"learning_rate":0.05,"margin_threshold":0.0}'},
   {'outer_fold': 4,
    'selected_structure_json': '{"centroids_per_class":2,"epochs":3,"learning_rate":0.1,"margin_threshold":0.1}'},
   {'outer_fold': 5,
    'selected_structure_json': '{"centroids_per_class":3,"epochs":1,"learning_rate":0.1,"margin_threshold":0.0}'}]},
 'hdc_regression': {'model': 'COMMON_ENCODER_READOUT_BASELINE',
  'dimension': 10000,
  'levels': 51,
  'feature_k': 50,
  'ridge_alpha': 0.01,

## 15. Missing-modality conditions and protocol separation

In [3]:
missing_contract = json.loads((PHASE09_ROOT / 'configs/phase09_missing_modality_contract.json').read_text(encoding='utf-8'))
{'primary_protocol': missing_contract['primary_protocol'], 'conditions': missing_contract['conditions'], 'new_training_runs': missing_contract['new_training_runs'], 'full_reference': missing_contract['full_primary_reference_policy'], 'sudden_test_time_missingness': missing_contract['sudden_test_time_missingness'], 'protocol_separation': missing_contract['protocol_separation']} 

{'primary_protocol': 'RETRAIN_WITHOUT_MODALITY',
 'conditions': [{'condition': 'FULL_PRIMARY_REFERENCE',
   'removed_modality': None,
   'removed_features': 0,
   'feature_count': 1176,
   'policy': 'REUSED_NOT_RETRAINED'},
  {'condition': 'MISSING_PHYSIOLOGICAL',
   'removed_modality': 'physiological_features',
   'removed_features': 233,
   'feature_count': 943,
   'policy': 'RETRAIN_WITHOUT_MODALITY'},
  {'condition': 'MISSING_EYE_TRACKING',
   'removed_modality': 'eye_tracking_features',
   'removed_features': 416,
   'feature_count': 760,
   'policy': 'RETRAIN_WITHOUT_MODALITY'},
  {'condition': 'MISSING_HEAD_MOVEMENT',
   'removed_modality': 'head_movement_features',
   'removed_features': 159,
   'feature_count': 1017,
   'policy': 'RETRAIN_WITHOUT_MODALITY'},
  {'condition': 'MISSING_FLIGHT_PARAMETER',
   'removed_modality': 'flight_parameter_features',
   'removed_features': 326,
   'feature_count': 850,
   'policy': 'RETRAIN_WITHOUT_MODALITY'},
  {'condition': 'MISSING_BODY_M

## 16. Deterministic LOSO assignments

In [4]:
loso_audit = json.loads((PHASE09_ROOT / 'audits/phase09_loso_assignment_audit.json').read_text(encoding='utf-8'))
{'status': loso_audit['status'], 'splits': loso_audit['splits'], 'assignment_rows': loso_audit['assignment_rows'], 'duplicate_run_keys': loso_audit['duplicate_run_keys'], 'checks': loso_audit['checks']} 

{'status': 'PASS',
 'splits': 35,
 'assignment_rows': 419,
 'duplicate_run_keys': 0,
 'checks': {'splits_35': True,
  'assignment_rows_419': True,
  'each_subject_held_out_once': True,
  'each_run_key_once': True,
  'no_run_key_missing': True,
  'training_subjects_34': True,
  'test_subjects_1': True,
  'config_selection_excludes_test_subject': True,
  'no_new_inner_cv': True}}

## 17. LOSO configuration mapping and leakage guard

In [5]:
mapping = json.loads((PHASE09_ROOT / 'configs/phase09_loso_config_mapping.json').read_text(encoding='utf-8'))
mapping_audit = json.loads((PHASE09_ROOT / 'audits/phase09_config_mapping_leakage_audit.json').read_text(encoding='utf-8'))
{'mapping_rule': mapping['mapping_rule'], 'mapped_subjects': len(mapping['mappings']), 'first_mapping': mapping['mappings'][0], 'last_mapping': mapping['mappings'][-1], 'leakage_audit': mapping_audit} 

{'mapping_rule': "Map each LOSO test subject to its Phase 03 original frozen outer_fold and reuse only that fold's upstream-selected configuration.",
 'mapped_subjects': 35,
 'first_mapping': {'loso_split': 1,
  'test_subject': 'sub-cp003',
  'original_outer_fold': 4,
  'train_subject_count': 34,
  'test_subject_count': 1,
  'test_subject_in_original_outer_training': False,
  'test_subject_excluded_from_original_config_selection_evidence': True,
  'new_inner_cv_authorized': False,
  'loso_test_subject_parameter_access': False,
  'fit_policy': 'Refit frozen configuration on the 34 LOSO training subjects only.',
  'config_sources': {'hdc_classification': {'path': 'E:\\hdc-vr-pilot\\experiments\\phase_06_hdc_variant_screening\\configs\\phase06_freeze.json',
    'locator': 'best_classification_hdc.fold_selected_structures[outer_fold=4]'},
   'hdc_regression': {'path': 'E:\\hdc-vr-pilot\\experiments\\phase_06_hdc_variant_screening\\configs\\phase06_freeze.json',
    'locator': 'best_regress

## 18. Dynamically enumerated 720-run execution matrix

In [6]:
execution = json.loads((PHASE09_ROOT / 'configs/phase09_execution_manifest.json').read_text(encoding='utf-8'))
run_audit = json.loads((PHASE09_ROOT / 'audits/phase09_run_matrix_audit.json').read_text(encoding='utf-8'))
{'training_run_count': execution['training_run_count'], 'duplicate_run_identifiers': execution['duplicate_run_identifiers'], 'by_protocol': execution['run_counts_by_protocol'], 'by_model': execution['run_counts_by_model'], 'first_run': execution['training_runs'][0], 'last_run': execution['training_runs'][-1], 'audit_status': run_audit['status'], 'all_statuses': sorted({item['status'] for item in execution['training_runs']})} 

{'training_run_count': 720,
 'duplicate_run_identifiers': 0,
 'by_protocol': {'RETRAIN_WITHOUT_MODALITY': 300,
  'LEAVE_ONE_SUBJECT_OUT': 420},
 'by_model': {'hdc_classification': 300,
  'hdc_regression': 300,
  'traditional_classification': 60,
  'traditional_regression': 60},
 'first_run': {'run_identifier': 'mm__missing_physiological__hdc_classification__fold_1__seed_42',
  'condition': 'MISSING_PHYSIOLOGICAL',
  'protocol': 'RETRAIN_WITHOUT_MODALITY',
  'task': 'classification',
  'model_key': 'hdc_classification',
  'model_family': 'HDC+OnlineHD Hybrid',
  'outer_fold': 1,
  'loso_subject': None,
  'seed_or_canonical': 42,
  'feature_count': 943,
  'config_source': {'path': 'E:\\hdc-vr-pilot\\experiments\\phase_06_hdc_variant_screening\\configs\\phase06_freeze.json',
   'locator': 'best_classification_hdc.fold_selected_structures[outer_fold=1]'},
  'expected_test_run_keys': ['sub-cp004_ses-20210330_level-01B_run-001',
   'sub-cp004_ses-20210330_level-01B_run-007',
   'sub-cp004_se

## 19. Read-only checkpoint portability conclusion

In [7]:
portability = json.loads((PHASE09_ROOT / 'audits/phase09_checkpoint_portability_audit.json').read_text(encoding='utf-8'))
{'audit_status': portability['status'], 'protocol_status': portability['protocol_status'], 'interfaces': {key: {'portable': value['portable'], 'checks': value['checks'], 'reproduction_attempted': value['reproduction_attempted']} for key, value in portability['interfaces'].items()}, 'training_executed': portability['training_executed'], 'predictions_generated': portability['predictions_generated']} 

{'audit_status': 'PASS',
 'protocol_status': 'NOT_FEASIBLE_DUE_TO_CHECKPOINT_INTERFACE',
 'interfaces': {'traditional_classification': {'portable': False,
   'checks': {'complete_loadable_model_saved': False,
    'training_fold_preprocessing_state_saved': False,
    'feature_order_saved': False,
    'hdc_encoder_state_saved_if_required': True,
    'full_input_prediction_reproduced_without_refit': False},
   'reproduction_attempted': False},
  'traditional_regression': {'portable': False,
   'checks': {'complete_loadable_model_saved': False,
    'training_fold_preprocessing_state_saved': False,
    'feature_order_saved': False,
    'hdc_encoder_state_saved_if_required': True,
    'full_input_prediction_reproduced_without_refit': False},
   'reproduction_attempted': False},
  'hdc_classification': {'portable': False,
   'checks': {'complete_loadable_model_saved': False,
    'training_fold_preprocessing_state_saved': False,
    'feature_order_saved': False,
    'hdc_encoder_state_saved_if

## 20. Generalization boundaries and statistical unit

In [8]:
guardrails = json.loads((PHASE09_ROOT / 'audits/phase09_generalization_guardrail_audit.json').read_text(encoding='utf-8'))
statistics = json.loads((PHASE09_ROOT / 'configs/phase09_statistical_rules.json').read_text(encoding='utf-8'))
{'guardrail_status': guardrails['status'], 'allowed': guardrails['allowed'], 'not_feasible': guardrails['not_feasible'], 'forbidden_scenario_proxies': guardrails['forbidden_scenario_proxies'], 'statistical_unit': statistics['statistical_unit'], 'formal_statistics_executed': statistics['formal_statistics_executed']} 

{'guardrail_status': 'PASS',
 'allowed': ['subject generalization via LOSO',
  'missing-modality robustness',
  'subject-level performance heterogeneity',
  'dependence on the flight modality'],
 'not_feasible': {'unseen_session': 'NOT_FEASIBLE_DUE_TO_METADATA',
  'unseen_scenario': 'NOT_FEASIBLE_DUE_TO_METADATA',
  'task_template': 'NOT_FEASIBLE_DUE_TO_METADATA',
  'route_configuration': 'NOT_FEASIBLE_DUE_TO_METADATA'},
 'forbidden_scenario_proxies': ['difficulty_level',
  'difficulty_level_raw',
  'target_class',
  'target_score',
  'run order',
  'feature clustering'],
 'statistical_unit': 'subject_id',
 'formal_statistics_executed': False}

## 21. Static Contract Freeze tests and audits

In [9]:
test_process = subprocess.run([sys.executable, '-m', 'unittest', 'discover', '-s', str(PHASE09_ROOT / 'tests'), '-p', 'test_phase09_contract.py', '-v'], cwd=PHASE09_ROOT, capture_output=True, text=True, check=False)
assert test_process.returncode == 0, test_process.stdout + test_process.stderr
audit_names = ['phase09_contract_freeze_audit.json', 'phase09_run_matrix_audit.json', 'phase09_loso_assignment_audit.json', 'phase09_config_mapping_leakage_audit.json', 'phase09_missing_modality_contract_audit.json', 'phase09_checkpoint_portability_audit.json', 'phase09_generalization_guardrail_audit.json']
audit_statuses = {name: json.loads((PHASE09_ROOT / 'audits' / name).read_text(encoding='utf-8'))['status'] for name in audit_names}
{'static_tests': 'PASS', 'audit_statuses': audit_statuses, 'details': (test_process.stdout + test_process.stderr)[-3000:], 'model_training_executed': False, 'predictions_generated': False} 

{'static_tests': 'PASS',
 'audit_statuses': {'phase09_contract_freeze_audit.json': 'PASS',
  'phase09_run_matrix_audit.json': 'PASS',
  'phase09_loso_assignment_audit.json': 'PASS',
  'phase09_config_mapping_leakage_audit.json': 'PASS',
  'phase09_missing_modality_contract_audit.json': 'PASS',
  'phase09_checkpoint_portability_audit.json': 'PASS',
  'phase09_generalization_guardrail_audit.json': 'PASS'},
 'details': 'test_checkpoint_portability_is_not_fabricated (test_phase09_contract.Phase09ContractTests.test_checkpoint_portability_is_not_fabricated) ... ok\ntest_dynamic_720_run_matrix (test_phase09_contract.Phase09ContractTests.test_dynamic_720_run_matrix) ... ok\ntest_initialization_and_upstream_gate (test_phase09_contract.Phase09ContractTests.test_initialization_and_upstream_gate) ... ok\ntest_loso_assignment_and_config_mapping (test_phase09_contract.Phase09ContractTests.test_loso_assignment_and_config_mapping) ... ok\ntest_missing_modality_contract (test_phase09_contract.Phase09Co

## 22. Final Contract Freeze persistence state

In [1]:
from pathlib import Path
import json
PHASE09_ROOT = Path.cwd().resolve()
freeze_config = json.loads((PHASE09_ROOT / 'configs/phase09_contract_freeze.json').read_text(encoding='utf-8'))
artifact_path = PHASE09_ROOT / 'audits/phase09_contract_artifact_audit.json'
artifact = json.loads(artifact_path.read_text(encoding='utf-8')) if artifact_path.exists() else {'status': 'PENDING_EXTERNAL_PERSISTENCE_AUDIT'}
{'phase09_status': freeze_config['status'], 'contract_artifact_audit': artifact['status'], 'authorized_training_runs': freeze_config['authorized_training_runs'], 'model_training_executed': freeze_config['training_executed'], 'predictions_generated': freeze_config['predictions_generated'], 'ready_for_execution': freeze_config['ready_for_execution']} 

{'phase09_status': 'CONTRACT_FROZEN_NOT_TRAINED',
 'contract_artifact_audit': 'PASS',
 'authorized_training_runs': 720,
 'model_training_executed': False,
 'predictions_generated': False,
 'ready_for_execution': True}

# Phase 09 Execution Evidence

This appended section records the authorized 720-run execution and raw prediction evidence. It performs no OOF consolidation, seed aggregation, formal statistics, or Phase 10 work.

## 23. Executor static validation

In [1]:
from pathlib import Path
import json, subprocess, sys
PHASE09_ROOT = Path.cwd().resolve()
sys.path.insert(0, str(PHASE09_ROOT / 'scripts'))
validation = json.loads((PHASE09_ROOT / 'audits/phase09_executor_validation_audit.json').read_text(encoding='utf-8'))
{'status': validation['status'], 'checks': validation['checks'], 'authorized_run_count': validation['dry_run_unique_runs'], 'training_executed': validation['training_executed']} 

{'status': 'PASS',
 'checks': {'exact_frozen_manifest_match': True,
  'unique_run_identifiers_720': True,
  'output_paths_no_collisions': True,
  'full_primary_reference_excluded': True,
  'sudden_test_time_missingness_excluded': True,
  'phase10_excluded': True,
  'performance_features_excluded': True,
  'expected_raw_rows_30168': True},
 'authorized_run_count': 720,
 'training_executed': False}

## 24. Exact 720-run dry-run summary

In [2]:
dry_run = validation
{'status': dry_run['status'], 'authorized_run_count': dry_run['dry_run_unique_runs'], 'expected_raw_prediction_rows': dry_run['expected_raw_prediction_rows'], 'exact_manifest_match': dry_run['checks']['exact_frozen_manifest_match'], 'training_executed': dry_run['training_executed']} 

{'status': 'PASS',
 'authorized_run_count': 720,
 'expected_raw_prediction_rows': 30168,
 'exact_manifest_match': True,
 'training_executed': False}

## 25. Missing-modality execution coverage

In [3]:
coverage = json.loads((PHASE09_ROOT / 'audits/phase09_execution_coverage_audit.json').read_text(encoding='utf-8'))
{'status': coverage['status'], 'completed_by_protocol': coverage['run_counts_by_protocol'], 'completed_conditions': coverage['missing_modality_conditions_completed'], 'checks': coverage['checks']} 

{'status': 'PASS',
 'completed_by_protocol': {'RETRAIN_WITHOUT_MODALITY': 300,
  'LEAVE_ONE_SUBJECT_OUT': 420},
 'completed_conditions': ['MISSING_BODY_MOVEMENT',
  'MISSING_EYE_TRACKING',
  'MISSING_FLIGHT_PARAMETER',
  'MISSING_HEAD_MOVEMENT',
  'MISSING_PHYSIOLOGICAL'],
 'checks': {'primary_checksum': True,
  'fold_checksum': True,
  'executor_validation': True,
  'runs_720': True,
  'raw_rows_30168': True,
  'duplicate_run_id_run_key_0': True,
  'invalid_runs_0': True,
  'schema_failures_0': True,
  'finite_predictions': True,
  'bounded_predictions': True,
  'upstream_files_modified_0': True,
  'frozen_contract_files_modified_0': True}}

## 26. LOSO execution coverage

In [4]:
{'status': coverage['status'], 'completed_subject_count': len(coverage['loso_subjects_completed']), 'completed_subjects': coverage['loso_subjects_completed'], 'loso_runs': coverage['run_counts_by_protocol']['LEAVE_ONE_SUBJECT_OUT']} 

{'status': 'PASS',
 'completed_subject_count': 35,
 'completed_subjects': ['sub-cp003',
  'sub-cp004',
  'sub-cp005',
  'sub-cp006',
  'sub-cp008',
  'sub-cp009',
  'sub-cp011',
  'sub-cp012',
  'sub-cp013',
  'sub-cp014',
  'sub-cp015',
  'sub-cp016',
  'sub-cp017',
  'sub-cp018',
  'sub-cp019',
  'sub-cp020',
  'sub-cp022',
  'sub-cp023',
  'sub-cp024',
  'sub-cp025',
  'sub-cp026',
  'sub-cp027',
  'sub-cp028',
  'sub-cp029',
  'sub-cp030',
  'sub-cp031',
  'sub-cp032',
  'sub-cp033',
  'sub-cp035',
  'sub-cp036',
  'sub-cp037',
  'sub-cp038',
  'sub-cp039',
  'sub-cp042',
  'sub-cp043'],
 'loso_runs': 420}

## 27. Checkpoint and recovery integrity

In [5]:
checkpoint = json.loads((PHASE09_ROOT / 'audits/phase09_checkpoint_integrity_audit.json').read_text(encoding='utf-8'))
{'status': checkpoint['status'], 'completed_checkpoints': checkpoint['completed_checkpoints'], 'expected_checkpoints': checkpoint['expected_checkpoints'], 'invalid_runs': checkpoint['invalid_runs'], 'load_or_hash_failures': checkpoint['load_or_hash_failures']} 

{'status': 'PASS',
 'completed_checkpoints': 720,
 'expected_checkpoints': 720,
 'invalid_runs': [],
 'load_or_hash_failures': []}

## 28. Raw prediction row evidence

In [6]:
{'status': coverage['status'], 'raw_prediction_rows': coverage['raw_prediction_rows'], 'expected_raw_prediction_rows': coverage['expected_raw_prediction_rows'], 'duplicate_run_id_run_key_pairs': coverage['duplicate_run_id_run_key_pairs'], 'completed_by_model': coverage['run_counts_by_model']} 

{'status': 'PASS',
 'raw_prediction_rows': {'total': 30168,
  'missing_modality': 25140,
  'loso': 5028,
  'hdc_classification': 12570,
  'hdc_regression': 12570,
  'traditional_classification': 2514,
  'traditional_regression': 2514},
 'expected_raw_prediction_rows': {'total': 30168,
  'missing_modality': 25140,
  'loso': 5028,
  'hdc_classification': 12570,
  'hdc_regression': 12570,
  'traditional_classification': 2514,
  'traditional_regression': 2514},
 'duplicate_run_id_run_key_pairs': 0,
 'completed_by_model': {'hdc_classification': 300,
  'hdc_regression': 300,
  'traditional_classification': 60,
  'traditional_regression': 60}}

## 29. Leakage and feature-exclusion audits

In [7]:
leakage = json.loads((PHASE09_ROOT / 'audits/phase09_execution_leakage_audit.json').read_text(encoding='utf-8'))
features = json.loads((PHASE09_ROOT / 'audits/phase09_feature_exclusion_audit.json').read_text(encoding='utf-8'))
{'leakage_status': leakage['status'], 'feature_exclusion_status': features['status'], 'leakage_failures': leakage['leakage_failures'], 'feature_failures': features['feature_failures']} 

{'leakage_status': 'PASS',
 'feature_exclusion_status': 'PASS',
 'leakage_failures': [],
 'feature_failures': []}

## 30. Execution artifact audit

In [8]:
artifacts = json.loads((PHASE09_ROOT / 'audits/phase09_execution_artifact_audit.json').read_text(encoding='utf-8'))
{'status': artifacts['status'], 'artifact_count': artifacts['artifact_count'], 'expected_artifact_count': artifacts['expected_artifact_count'], 'missing_or_invalid_runs': artifacts['missing_or_invalid_runs'], 'upstream_files_modified': artifacts['upstream_files_modified'], 'frozen_contract_files_modified': artifacts['frozen_contract_files_modified']} 

{'status': 'PASS',
 'artifact_count': 3600,
 'expected_artifact_count': 3600,
 'missing_or_invalid_runs': [],
 'upstream_files_modified': [],
 'frozen_contract_files_modified': []}

## 31. Execution limitations and stopping boundary

In [9]:
limitations = {'full_primary_reference_retrained': False, 'sudden_test_time_missingness': 'NOT_FEASIBLE_DUE_TO_CHECKPOINT_INTERFACE', 'oof_consolidation_executed': False, 'seed_aggregation_executed': False, 'formal_statistics_executed': False, 'phase10_started': False, 'phase09_final_freeze_executed': False}
limitations

{'full_primary_reference_retrained': False,
 'sudden_test_time_missingness': 'NOT_FEASIBLE_DUE_TO_CHECKPOINT_INTERFACE',
 'oof_consolidation_executed': False,
 'seed_aggregation_executed': False,
 'formal_statistics_executed': False,
 'phase10_started': False,
 'phase09_final_freeze_executed': False}

## 32. Final execution persistence state

In [1]:
from pathlib import Path
import json
PHASE09_ROOT = Path.cwd().resolve()
manifest = json.loads((PHASE09_ROOT / 'configs/phase09_execution_manifest.json').read_text(encoding='utf-8'))
persistence_path = PHASE09_ROOT / 'audits/phase09_execution_notebook_persistence_audit.json'
persistence = json.loads(persistence_path.read_text(encoding='utf-8')) if persistence_path.exists() else {'status': 'PENDING_EXTERNAL_PERSISTENCE_AUDIT'}
{'phase09_execution_status': manifest['status'], 'completed_training_runs': manifest['completed_training_runs'], 'raw_prediction_rows': manifest['raw_prediction_rows'], 'notebook_persistence': persistence['status'], 'ready_for_oof_consolidation': manifest.get('ready_for_oof_consolidation', False)} 

{'phase09_execution_status': 'EXECUTION_COMPLETE_PENDING_OOF_CONSOLIDATION',
 'completed_training_runs': 720,
 'raw_prediction_rows': 30168,
 'notebook_persistence': 'PASS',
 'ready_for_oof_consolidation': True}

## Phase 09 Final Analysis Evidence

Persisted presentation of canonical OOF, independently recalculated metrics, subject-level statistics, figures, and claim boundaries. No training, raw-prediction regeneration, freeze, or Phase 10 execution occurs here.

In [1]:
from pathlib import Path
import json, pandas as pd
PHASE09_ROOT=Path.cwd().resolve()
def audit(name): return json.loads((PHASE09_ROOT/'audits'/name).read_text(encoding='utf-8'))
index=pd.read_csv(PHASE09_ROOT/'results/oof/phase09_canonical_oof_index.csv')
refs=pd.read_csv(PHASE09_ROOT/'results/oof/phase09_full_primary_reference_index.csv')
{'canonical_rows':len(index),'reference_rows':len(refs),'coverage':audit('phase09_oof_coverage_audit.json')['status'],'alignment':audit('phase09_oof_alignment_audit.json')['status'],'leakage':audit('phase09_oof_leakage_audit.json')['status']}

{'canonical_rows': 10056,
 'reference_rows': 1676,
 'coverage': 'PASS',
 'alignment': 'PASS',
 'leakage': 'PASS'}

### Missing-modality robustness and subject-level inference

In [2]:
robustness=pd.read_csv(PHASE09_ROOT/'results/summaries/phase09_missing_modality_robustness.csv')
stats=pd.read_csv(PHASE09_ROOT/'results/summaries/phase09_pairwise_statistics.csv')
{'comparisons':len(stats),'subjects_per_comparison':sorted(stats.n_subjects.unique().tolist()),'holm_complete':bool(stats.p_value_holm.notna().all()),'bootstrap_resamples':sorted(stats.bootstrap_resamples.unique().tolist()),'largest_degradation_by_model':robustness.loc[robustness.groupby('model_key').mean_subject_degradation.idxmax(),['model_key','condition','mean_subject_degradation']].to_dict('records')}

{'comparisons': 20,
 'subjects_per_comparison': [35],
 'holm_complete': True,
 'bootstrap_resamples': [2000],
 'largest_degradation_by_model': [{'model_key': 'hdc_classification',
   'condition': 'MISSING_FLIGHT_PARAMETER',
   'mean_subject_degradation': 0.5180918288061145},
  {'model_key': 'hdc_regression',
   'condition': 'MISSING_FLIGHT_PARAMETER',
   'mean_subject_degradation': 0.5906271611173007},
  {'model_key': 'traditional_classification',
   'condition': 'MISSING_FLIGHT_PARAMETER',
   'mean_subject_degradation': 0.5102618413332699},
  {'model_key': 'traditional_regression',
   'condition': 'MISSING_FLIGHT_PARAMETER',
   'mean_subject_degradation': 0.718437785797222}]}

### LOSO stability, flight dependence, and generalization boundary

In [3]:
loso=pd.read_csv(PHASE09_ROOT/'results/summaries/phase09_loso_subject_stability.csv')
flight=pd.read_csv(PHASE09_ROOT/'results/summaries/phase09_flight_dependence_evidence.csv')
{'loso_subjects':sorted(loso.subjects.unique().tolist()),'flight_dependence_rows':len(flight),'flight_generalizable_behavior_claim':sorted(flight.generalizable_behavior_claim.unique().tolist()),'unseen_session':'NOT_FEASIBLE_DUE_TO_METADATA','unseen_scenario':'NOT_FEASIBLE_DUE_TO_METADATA','task_template':'NOT_FEASIBLE_DUE_TO_METADATA','route_configuration':'NOT_FEASIBLE_DUE_TO_METADATA'}

{'loso_subjects': [35],
 'flight_dependence_rows': 4,
 'flight_generalizable_behavior_claim': ['INCONCLUSIVE_DUE_TO_METADATA'],
 'unseen_session': 'NOT_FEASIBLE_DUE_TO_METADATA',
 'unseen_scenario': 'NOT_FEASIBLE_DUE_TO_METADATA',
 'task_template': 'NOT_FEASIBLE_DUE_TO_METADATA',
 'route_configuration': 'NOT_FEASIBLE_DUE_TO_METADATA'}

### Figures, reports, and pre-persistence verification

In [4]:
figures=sorted(p.name for p in (PHASE09_ROOT/'figures').glob('phase09_*.*'))
reports=sorted(str(p.relative_to(PHASE09_ROOT)) for p in (PHASE09_ROOT/'reports').rglob('*.md'))
verification=audit('phase09_final_analysis_verification.json')
{'figure_files':figures,'report_files':reports,'pre_persistence_status':verification['status'],'model_retraining_executed':False,'raw_predictions_regenerated':False,'phase09_freeze_executed':False,'phase10_executed':False}

{'figure_files': ['phase09_loso_stability_distribution.pdf',
  'phase09_loso_stability_distribution.png',
  'phase09_loso_subject_classification.pdf',
  'phase09_loso_subject_classification.png',
  'phase09_loso_subject_regression.pdf',
  'phase09_loso_subject_regression.png',
  'phase09_missing_modality_classification_curve.pdf',
  'phase09_missing_modality_classification_curve.png',
  'phase09_missing_modality_model_comparison.pdf',
  'phase09_missing_modality_model_comparison.png',
  'phase09_missing_modality_regression_curve.pdf',
  'phase09_missing_modality_regression_curve.png'],
 'report_files': ['reports\\analysis-output\\analysis-report.md',
  'reports\\analysis-output\\figure-catalog.md',
  'reports\\analysis-output\\stats-appendix.md',
  'reports\\phase09_final_analysis.md',
  'reports\\phase09_generalization_boundaries.md',
  'reports\\phase09_loso_stability_report.md',
  'reports\\phase09_missing_modality_report.md',
  'reports\\phase09_statistical_appendix.md'],
 'pre_per

## Final Analysis Takeaways

Phase 09 contains 10,056 canonical OOF rows plus 1,676 frozen Full Primary reference rows. Missing-modality effects are paired at the subject level (n=35), corrected with Holm within each preregistered model-task family, and accompanied by 2,000-resample bootstrap intervals. LOSO supports held-out-subject evaluation only. Flight generalizable behavior is inconclusive because session, scenario, task-template, and route metadata are unavailable. Analysis is complete pending a separate Phase 09 freeze; Phase 10 was not executed.

In [1]:
from pathlib import Path
import json
PHASE09_ROOT=Path.cwd().resolve()
manifest=json.loads((PHASE09_ROOT/'configs/phase09_execution_manifest.json').read_text(encoding='utf-8'))
{'phase09_status':manifest['status'],'ready_for_phase09_freeze':manifest.get('ready_for_phase09_freeze',False),'phase09_freeze_executed':manifest.get('phase09_freeze_executed',False),'phase10_executed':manifest.get('phase10_executed',False)}

{'phase09_status': 'ANALYSIS_COMPLETE_PENDING_FREEZE',
 'ready_for_phase09_freeze': True,
 'phase09_freeze_executed': False,
 'phase10_executed': False}

## Phase 09 Final Freeze Summary

Executed final-freeze evidence. All historical cells and outputs above are preserved; no model training, prediction regeneration, or Phase 10 execution occurs here.

In [1]:
from pathlib import Path
import json
root=Path.cwd().resolve()
freeze=json.loads((root/'configs/phase09_freeze.json').read_text(encoding='utf-8'))
manifest=json.loads((root/'manifests/phase09_final_manifest.json').read_text(encoding='utf-8'))
{'runs':'720/720','raw_prediction_rows':30168,'canonical_oof_rows':10056,'loso_subjects':'35/35','missing_modality_analysis':'COMPLETE','subject_level_statistics':'COMPLETE','final_audits':'PASS','final_manifest_saved':True,'phase09_freeze_saved':True,'phase09_status':freeze['status'],'phase10_executed':freeze['phase10_executed'],'ready_for_phase10':freeze['ready_to_proceed_to_phase10']}

{'runs': '720/720',
 'raw_prediction_rows': 30168,
 'canonical_oof_rows': 10056,
 'loso_subjects': '35/35',
 'missing_modality_analysis': 'COMPLETE',
 'subject_level_statistics': 'COMPLETE',
 'final_audits': 'PASS',
 'final_manifest_saved': True,
 'phase09_freeze_saved': True,
 'phase09_status': 'FROZEN',
 'phase10_executed': False,
 'ready_for_phase10': True}